extra/crc32-x86-pclmul/crc_intel_pclmul.c

```diff
@@ -508,4 +510,41 @@ crc32_intel_pclmul (u32 *pcrc, const byte *inbuf, size_t inlen)
 #endif
 }
 
+static inline uint32_t cpuid(uint32_t *ecx, uint32_t *edx)
+{
+  uint32_t level;
+
+  asm("cpuid" : "=a"(level) : "a"(0) : "ebx", "ecx", "edx");
+
+  if (level < 1)
+  {
+    return level;
+  }
+
+  asm("cpuid" : "=c"(*ecx), "=d"(*edx) : "a"(1) : "ebx");
+
+  return level;
+}
+
+int is_pclmul_enabled()
+{
+  int pclmul_enabled= 0;
+  uint32_t ecx, edx;
+
+  if (cpuid(&ecx, &edx) > 0)
+  {
+    pclmul_enabled= ((ecx >> 19) & 1) && ((ecx >> 1) & 1);
+  }
+
+  return (pclmul_enabled);
+}
```

Comment:
The parameter `edx` parameter is redundant and should be removed. I would merge the functions `is_pclmul_enabled()` and `cpuid()` as follows, and explicitly specify an empty parameter list, since this is C and not C++ code:
```c
#ifdef __GNUC__
int crc32_pclmul_enabled(void)
{
  int eax,ecx;
  /* We assume that the CPUID instruction and its parameter 1 are available.
  We do not support any precursors of the Intel 80486. */
  asm ("cpuid" : "=a"(eax), "=c"(ecx) : "0"(1) : "ebx", "edx");
  return !(~ecx & (1 << 19 | 1 << 1));
}
#elif defined _MSC_VER
#include <intrin.h>
int crc32_pclmul_enabled(void)
{
  /* We assume that the CPUID instruction and its parameter 1 are available.
  We do not support any precursors of the Intel 80486. */
  int regs[4];
  __cpuid(regs, 1);
  return !(~regs[2] & (1 << 19 | 1 << 1));
}
#else
int crc32_pclmul_enabled(void)
{
  return 0;
}
#endif
```
Remember that also other architectures exist, such as the s390x.

Reply:
afdd6b1da102d5cef2dcc24ea0636381285d7899 cleaned up the `ut_crc32_init()`. While checking the results, I made the surprising discovery that on buildbot.askmonty.org, the `cpuid` check always fails (also for 10.2) and the non-accelerated implementation is chosen on AMD64 even though the VM hosts do support SSE4.2. I suspect that it is due to something in the configuration virtual machine environment. When I downloaded and disassembled a binary package, the code looked correct to me.

Reply:
replaced as recommended.